org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -2703,6 +2777,175 @@ class MarkedString {
 	}
 }
 
+interface WorkDoneProgressNotification {
+	def WorkDoneProgressKind getKind()
+}
+ 
+
+@JsonRpcData
+class WorkDoneProgressBegin implements WorkDoneProgressNotification {
+
+	override WorkDoneProgressKind getKind() {
+		WorkDoneProgressKind.BEGIN
+	}
+
+	/**
+	 * Mandatory title of the progress operation. Used to briefly inform about
+	 * the kind of operation being performed.
+	 *
+	 * Examples: "Indexing" or "Linking dependencies".
+	 */
+	 @NonNull
+	 String title
+
+	/**
+	 * Controls if a cancel button should show to allow the user to cancel the
+	 * long running operation. Clients that don't support cancellation are allowed
+	 * to ignore the setting.
+	 */
+	Boolean cancellable
+
+	/**
+	 * Optional, more detailed associated progress message. Contains
+	 * complementary information to the `title`.
+	 *
+	 * Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
+	 * If unset, the previous progress message (if any) is still valid.
+	 */
+	String message
+
+	/**
+	 * Optional progress percentage to display (value 100 is considered 100%).
+	 * If not provided infinite progress is assumed and clients are allowed
+	 * to ignore the `percentage` value in subsequent in report notifications.
+	 *
+	 * The value should be steadily rising. Clients are free to ignore values
+	 * that are not following this rule.
+	 */
+	Integer percentage;
+}
+
+@JsonRpcData
+class WorkDoneProgressReport implements WorkDoneProgressNotification {
+
+	override WorkDoneProgressKind getKind() {
+		WorkDoneProgressKind.REPORT
+	}
+
+	/**
+	 * Controls enablement state of a cancel button. This property is only valid if a cancel
+	 * button got requested in the `WorkDoneProgressStart` payload.
+	 *
+	 * Clients that don't support cancellation or don't support control the button's
+	 * enablement state are allowed to ignore the setting.
+	 */
+	Boolean cancellable
+
+	/**
+	 * Optional, more detailed associated progress message. Contains
+	 * complementary information to the `title`.
+	 *
+	 * Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
+	 * If unset, the previous progress message (if any) is still valid.
+	 */
+	String message
+
+	/**
+	 * Optional progress percentage to display (value 100 is considered 100%).
+	 * If not provided infinite progress is assumed and clients are allowed
+	 * to ignore the `percentage` value in subsequent in report notifications.
+	 *
+	 * The value should be steadily rising. Clients are free to ignore values
+	 * that are not following this rule.
+	 */
+	Integer percentage;
+}
+
+@JsonRpcData
+class WorkDoneProgressEnd implements WorkDoneProgressNotification {
+
+	override WorkDoneProgressKind getKind() {
+		WorkDoneProgressKind.END
+	}
+
+	/**
+	 * Optional, a final message indicating to for example indicate the outcome
+	 * of the operation.
+	 */
+	String message
+}
+
+class ProgressToken extends Either<String, Number> {
```

Comment:
As stated above, I don't think this carries its own weight. Also extending `Either` does not seem to be a good choice. One would usually also not extend `ArrayList` or `HashMap` and I picture `Either` to be a data type that plays in the same ballpark as the types from java.util.

Reply:
I see this differently. Supporting the creation of Either subclasses via a public constructor is a small change. Using the type `ProgressToken` in a field makes the protocol much better readable compared to `Either<String, Number>`.

However, if we can reduce the references to this type to only a few via an abstract superclass as you proposed above, the situation would be different.

Reply:
Sorry i don't catch your last sentence.

> However, if we can reduce the references to this type to only a few via an abstract superclass as you proposed above, the situation would be different.

What is the proposition ? Create an abstract class common between Either and ProgressToken ?

Reply:
I meant this proposal: https://github.com/eclipse/lsp4j/pull/442#discussion_r476240366